sn_networking/src/replication_fetcher.rs

```diff
@@ -164,15 +167,20 @@ impl ReplicationFetcher {
                         "Prune record {:?} at {holder:?} from the replication_fetcher due to timeout.",
                         PrettyPrintRecordKey::from(key)
                     );
+                    failed_holders.push(*holder);
                     true
                 } else {
                     false
                 }
             } else {
                 false
             };
-            is_fetching.is_none() || !is_expired
+
+            // if the node isn't known to be slow or the fetching is not expired we keep the entry
+            !failed_holders.contains(holder) && ( is_fetching.is_none() || !is_expired )
```

Comment:
failed_holders will only be filled when hit an expired entry,
which means: for those entries in the `to_be_fetched` before the `expired` one, they will be retained as at that moment, the `node in question` is not in the `black list` yet.